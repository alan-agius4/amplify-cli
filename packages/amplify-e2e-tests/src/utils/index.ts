import { join } from 'path';
import * as os from 'os';
import { mkdirSync } from 'fs';
import * as rimraf from 'rimraf';
import * as fs from 'fs-extra';
import { config } from 'dotenv';
export * from './projectMeta';
export * from './transformConfig';
export * from './awsExports';
export * from './sdk-calls';
export * from './api';
import * as nanoid from 'nanoid';
import { Readable } from 'stream';

const NANOID_LENGTH = 7;
const RADIX = 10;
// run dotenv config to update env variable
config();

export function getCLIPath() {
  return join(__dirname, '..', '..', '..', 'amplify-cli', 'bin', 'amplify');
}

export function createNewProjectDir(name: string, inCIMode: Boolean = isCI()): string {
  // if (!root) {
  //   root = join(__dirname, '../../../..', `amplify-integ-${Math.round(Math.random() * 100)}-test-${Math.round(Math.random() * 1000)}`);
  // }
  const dir = inCIMode ? join(os.homedir(), 'amplify-cli-e2e-tests') : join(__dirname, '../../../..');
  const root = join(dir, `${name}-${nanoid(NANOID_LENGTH)}`);
  mkdirSync(root);
  return root;
}

export function deleteProjectDir(root: string, inCIMode = isCI()) {
  if (!inCIMode) {
    return rimraf.sync(root);
  }
}

export function isCI(): Boolean {
  return process.env.CI ? true : false;
}

export function getEnvVars(): {} {
  return { ...process.env };
}

export function writeStdOutToDisk(filename: string, directoryName: string, stream: Readable) {
  const fileStartingWithNumberRegExp = /^(\d+)[-].*/;
  let fileNumber: number = 0;
  const logDirectoryName = join(directoryName, 'logs');
  fs.ensureDirSync(logDirectoryName);
  try {
    const filesWithNumber = fs
      .readdirSync(logDirectoryName)
      .filter(f => fs.statSync(join(logDirectoryName, f)).isFile())
      .filter(f => fileStartingWithNumberRegExp.test(f))
      .sort();
    if (filesWithNumber.length) {
      const lastIndex = fileStartingWithNumberRegExp.exec(filesWithNumber.pop())[1];
      fileNumber = Number.parseInt(lastIndex, RADIX);
    }
    fs.writeFileSync(join(logDirectoryName, `${fileNumber}-${filename}`), stream.);
  } catch (e) {
    console.log(`Error when writing content to disk ${join(directoryName, filename)}`, e);
  }
}
