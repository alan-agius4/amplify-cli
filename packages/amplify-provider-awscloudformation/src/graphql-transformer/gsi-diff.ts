import { DynamoDB, IntrinsicFunction } from 'cloudform';
import * as _ from 'lodash';
import { GlobalSecondaryIndex } from 'cloudform-types/types/dynamoDb/table';
import Table from 'cloudform-types/types/glue/table';
import { diff as getDiffs } from 'deep-diff';

export enum GSIChange {
  Add = 'ADD',
  Update = 'UPDATE',
  Delete = 'DELETE',
}
export type IndexChange = {
  type: GSIChange;
  indexName: string;
};

export const getGSIDiffs = (current: DynamoDB.Table, next: DynamoDB.Table) => {
  if (
    current.Properties.GlobalSecondaryIndexes instanceof IntrinsicFunction ||
    next.Properties.GlobalSecondaryIndexes instanceof IntrinsicFunction
  ) {
    return [];
  }
  const currentIndexes = current.Properties.GlobalSecondaryIndexes ?? [];
  const nextIndexes = next.Properties.GlobalSecondaryIndexes;
  return handleIndividualDiff(currentIndexes, nextIndexes);
};

export const handleIndividualDiff = (currentIndexes: GlobalSecondaryIndex[], nextIndexes: GlobalSecondaryIndex[]): IndexChange[] => {
  // Create  Record<IndexName, Index>
  const currentIndexByIndexName = _.keyBy(currentIndexes, 'IndexName');
  // create an array of indexes
  const currentIndexNames = Object.keys(currentIndexByIndexName);

  // Create  Record<IndexName, Index>
  const nextIndexByIndexName = _.keyBy(nextIndexes, 'IndexName');
  // create an array of indexes
  const nextIndexNames = Object.keys(nextIndexByIndexName);

  // Get the indexes which are not in both current and next indexes
  const addedOrRemovedIndexNames = _.xor(currentIndexNames, nextIndexNames);

  // Partition them as added/removed indexes
  const [indexToRemove, indexToAdd] = _.partition(addedOrRemovedIndexNames, indexName => currentIndexNames.includes(indexName));

  // Get all the indexes that are in bot current and next indexes
  const possiblyModifiedIndexNames = _.xor([...currentIndexNames, ...nextIndexNames], addedOrRemovedIndexNames);

  const modifiedIndexes = possiblyModifiedIndexNames
    .map(indexName => {
      return handlerGSIUpdate(currentIndexByIndexName[indexName], nextIndexByIndexName[indexName]);
    })
    .filter(change => Boolean(change));

  return [
    ...indexToRemove.map(idx => ({
      type: GSIChange.Delete,
      indexName: idx,
    })),
    ...indexToAdd.map(idx => ({
      type: GSIChange.Add,
      indexName: idx,
    })),
    ...modifiedIndexes,
  ];
};

export const handlerGSIUpdate = (currentIndex: GlobalSecondaryIndex, nextIndex: GlobalSecondaryIndex): IndexChange | undefined => {
  const diffs = getDiffs(currentIndex, nextIndex);
  if (currentIndex.IndexName instanceof IntrinsicFunction) {
    return;
  }
  const isModeifed = diffs?.some(diff => {
    const leaf = diff.path?.slice(-1)[0];
    return [
      'IndexName',
      'KeySchema',
      'AttributeName',
      'AttributeType',
      'KeyType',
      'NonKeyAttributes',
      'Projection',
      'ProjectionType',
    ].includes(leaf);
  });
  return isModeifed ? { type: GSIChange.Update, indexName: currentIndex.IndexName } : undefined;
};
