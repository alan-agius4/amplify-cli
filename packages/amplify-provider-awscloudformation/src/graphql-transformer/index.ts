export { GraphQLResourceManager } from './amplify-graphql-resource-mananger';
