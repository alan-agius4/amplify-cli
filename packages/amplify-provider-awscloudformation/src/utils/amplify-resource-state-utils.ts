import { Template } from 'cloudform-types';
import { GlobalSecondaryIndex, AttributeDefinition } from 'cloudform-types/types/dynamoDb/table';
import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import { CloudFormation } from 'aws-sdk';

export interface GSIRecord {
  attributeDefinition: AttributeDefinition[];
  gsi: GlobalSecondaryIndex;
}

export enum GSIStatus {
  add = 'add',
  edit = 'edit',
  delete = 'delete',
  batchAdd = 'batchAdd',
  batchDelete = 'batchDelete',
  none = 'none',
}

export const getStackParameters = async(cfnClient: CloudFormation, StackId: string): Promise<any> => {
  const apiStackInfo = await cfnClient
    .describeStacks({
      StackName: StackId,
    }).promise();
  return apiStackInfo.Stacks[0].Parameters.reduce((acc, param) => {
    acc[param.ParameterKey] = param.ParameterValue;
    return acc;
  }, {});
}

export const getTableARNS = async (cfnClient: CloudFormation, tables: string[], StackId: string): Promise<Map<string, string>> => {
  const arnMap: Map<string, string> = new Map();
  const apiResources = await cfnClient
    .describeStackResources({
      StackName: StackId,
    })
    .promise();
  for (const resource of apiResources.StackResources) {
    if (tables.includes(resource.LogicalResourceId)) {
      const tableStack = await cfnClient
        .describeStacks({
          StackName: resource.PhysicalResourceId,
        })
        .promise();
      const tableARN = tableStack.Stacks[0].Outputs.reduce((acc, out) => {
        if (out.OutputKey === `GetAtt${resource.LogicalResourceId}TableName`) {
          acc.push(out.OutputValue);
        }
        return acc;
      }, []);
      arnMap.set(resource.LogicalResourceId, tableARN[0]);
    }
  }
  return arnMap;
};
export class TemplateState {
  private changes: { [key: string]: string[] } = {};

  public has(key: string) {
    return Boolean(key in this.changes);
  }

  public isEmpty(): Boolean {
    return !Object.keys(this.changes).length;
  }

  public get(key: string): string[] {
    return this.changes[key];
  }

  public getLatest(key: string): Template | null {
    if (this.changes[key]) {
      const length = this.changes[key].length;
      return length ? JSON.parse(this.changes[key][length - 1]) : null;
    }
    return null;
  }

  public pop(key: string): Template {
    const template = this.changes[key].shift();
    if (_.isEmpty(this.changes[key])) {
      delete this.changes[key];
    }
    return JSON.parse(template);
  }

  public add(key: string, val: string): void {
    if (!(key in this.changes)) {
      this.changes[key] = [];
    }
    this.changes[key].push(val);
  }

  public getChangeCount(key: string): number {
    return this.changes[key].length;
  }

  public getKeys(): Array<string> {
    return Object.keys(this.changes);
  }
}
