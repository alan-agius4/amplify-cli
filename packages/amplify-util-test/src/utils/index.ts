
export { addCleanupTask } from './cleanup-task'
export async function getAmplifyMeta(context: any) {
  const amplifyMetaFilePath = context.amplify.pathManager.getAmplifyMetaFilePath();
  return context.amplify.readJsonFile(amplifyMetaFilePath);
}
