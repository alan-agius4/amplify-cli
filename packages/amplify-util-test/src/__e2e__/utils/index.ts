import { AmplifyAppSyncSimulator } from 'amplify-appsync-simulator';
import { ensureDynamoDBTables, configureDDBDataSource } from '../../utils/ddb-utils';
import { processResources } from '../../CFNParser/resource-processor';
import * as dynamoEmulator from '@conduitvc/dynamodb-emulator';
import * as fs from 'fs-extra';
import * as path from 'path';

export async function launchDDBLocal() {
    const dbPath = path.join(
        '/tmp',
        `amplify-cli-emulator-dynamodb-${Math.floor(Math.random() * 100)}`
    );
    fs.ensureDirSync(dbPath);
    const emulator = await dynamoEmulator.launch({
        dbPath,
        port: null
    });
    const client = await dynamoEmulator.getClient(emulator);
    return { emulator, dbPath, client };
}

export async function deploy(transformerOutput: any, client) {
    const stacks = Object.values(transformerOutput.stacks).reduce(
        (prev, stack: any) => {
            return { ...prev, ...stack.Resources };
        },
        { ...transformerOutput.rootStack.Resources }
    );

    let config = processResources(stacks, transformerOutput);
    await ensureDynamoDBTables(client, config);
    config = configureDDBDataSource(config, client.config);
    const simulator = await runAppSyncSimulator(config);
    return {simulator, config};
}

export async function terminateDDB(emulator, dbPath) {
    await emulator.terminate();
    fs.removeSync(dbPath);
}

export async function runAppSyncSimulator(
    config,
    port?: number,
    wsPort?: number
) {
    const appsyncSimulator = new AmplifyAppSyncSimulator(config, { port, wsPort });
    await appsyncSimulator.start();
    return appsyncSimulator;
}