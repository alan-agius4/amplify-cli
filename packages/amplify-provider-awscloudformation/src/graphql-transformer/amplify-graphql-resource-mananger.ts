import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import configurationManager from '../configuration-manager';
import { diff as getDiffs, Diff } from 'deep-diff';
import { sanityCheck } from 'graphql-transformer-core';
import { Template, DynamoDB } from 'cloudform-types';
import { $TSContext } from 'amplify-cli-core';
import { CloudFormation } from 'aws-sdk';
import { getStackParameters, GSIStatus, GSIRecord, TemplateState, getTableARNS } from '../utils/amplify-resource-state-utils';
import { GlobalSecondaryIndex, KeySchema, AttributeDefinition } from 'cloudform-types/types/dynamoDb/table';
import { DeploymentStep } from '../iterative-deployment/state-machine';
import { hashDirectory, ROOT_APPSYNC_S3_KEY } from '../upload-appsync-files';
import { DiffChanges, getGQLDiff, DiffableProject } from './utils';

export type GQLResourceManagerProps = {
  cfnClient: CloudFormation;
  resourceMeta: $ResourceMeta | null;
  backendDir: string;
  cloudBackendDir: string;
};

export type $ResourceMeta = {
  category: string;
  providerPlugin: string;
  resourceName: string;
  service: string;
  output: any;
  providerMetadata: {
    s3TemplateURL: string;
    logicalId: string;
  };
  stackId: string;
  DeploymentBucketName: string;
  [key: string]: any;
};

// TODO: Add unit testing
export class GraphQLResourceManager {
  static serviceName: string = 'AppSync';
  static categoryName: string = 'api';
  private cfnClient: CloudFormation;
  private resourceMeta: $ResourceMeta;
  private cloudBackendDir: string;
  private backendDir: string;
  private templateState: TemplateState;

  public static createInstance = async (context: $TSContext, gqlResource: any, StackId: string) => {
    try {
      const cred = await configurationManager.loadConfiguration(context);
      const cfn = new CloudFormation(cred);
      const apiStack = await cfn
        .describeStackResources({ StackName: StackId, LogicalResourceId: gqlResource.providerMetadata.logicalId })
        .promise();
      return new GraphQLResourceManager({
        cfnClient: cfn,
        resourceMeta: { ...gqlResource, stackId: apiStack.StackResources[0].PhysicalResourceId },
        backendDir: context.amplify.pathManager.getBackendDirPath(),
        cloudBackendDir: context.amplify.pathManager.getCurrentCloudBackendDirPath(),
      });
    } catch (err) {
      throw err;
    }
  };

  constructor(props: GQLResourceManagerProps) {
    if (!props.resourceMeta) {
      throw Error('No GraphQL API enabled.');
    }
    this.cfnClient = props.cfnClient;
    this.resourceMeta = props.resourceMeta;
    this.backendDir = path.join(props.backendDir, GraphQLResourceManager.categoryName, this.resourceMeta.resourceName);
    this.cloudBackendDir = path.join(props.cloudBackendDir, GraphQLResourceManager.categoryName, this.resourceMeta.resourceName);
    this.templateState = new TemplateState();
  }

  run = async (): Promise<DeploymentStep[]> | null => {
    const gqlDiff = getGQLDiff(this.backendDir, this.cloudBackendDir);
    try {
      sanityCheck(gqlDiff.diff, gqlDiff.current, gqlDiff.next);
    } catch (err) {
      if (err.name !== 'InvalidGSIMigrationError') {
        throw err;
      }
    }
    this.gsiManagement(gqlDiff.diff, gqlDiff.current, gqlDiff.next);
    return await this.getDeploymentSteps();
  };

  // save states to build with a copy of build on every deploy
  getDeploymentSteps = async (): Promise<DeploymentStep[]> => {
    let count = 0;
    const gqlSteps = new Array<DeploymentStep>();
    const tempDir = path.join(this.cloudBackendDir, 'build', 'states');
    const stateFileDir = path.join(this.backendDir, 'build');
    const tableArnMap = await getTableARNS(this.cfnClient, this.templateState.getKeys(), this.resourceMeta.stackId);
    const parameters = await getStackParameters(this.cfnClient, this.resourceMeta.stackId);
    const buildHash = await hashDirectory(this.backendDir);
    fs.ensureDirSync(tempDir);
    while (!this.templateState.isEmpty()) {
      fs.copySync(stateFileDir, path.join(tempDir, `${count}`));
      const tables = this.templateState.getKeys();
      const tableArns = [];
      tables.forEach(key => {
        tableArns.push(tableArnMap.get(key));
        const filepath = path.join(stateFileDir, `${count}`, 'stacks', `${key}.json`);
        fs.writeFileSync(filepath, JSON.stringify(this.templateState.pop(key), null, 2));
      });
      gqlSteps.push({
        stackTemplatePath: this.resourceMeta.providerMetadata.s3TemplateURL,
        parameters: { ...parameters, S3DeploymentRootKey: `${ROOT_APPSYNC_S3_KEY}/${buildHash}/states/${count}` },
        stackName: this.resourceMeta.stackId,
        tableNames: tableArns,
      });
      count++;
    }
    fs.moveSync(tempDir, path.join(stateFileDir, 'states'));
    return gqlSteps;
  };

  private gsiManagement = (diffs: DiffChanges<DiffableProject>, currentState: DiffableProject, nextState: DiffableProject) => {
    const gsiChanges = _.filter(diffs, diff => {
      return _.includes(diff.path, 'GlobalSecondaryIndexes');
    });
    for (const gsiChange of gsiChanges) {
      const tableName = gsiChange.path[3];
      const stackName = gsiChange.path[1].split('.')[0];
      const gsiStatus = this.gsiChangeStatus(gsiChange, currentState, nextState);
      const ddbResource = this.templateState.getLatest(stackName) || this.getStack(stackName, currentState);

      if (gsiStatus === GSIStatus.add) {
        const indexName = (gsiChange as any).item.rhs.IndexName;
        let gsiRecord = this.getGSIRecord(indexName, this.getTable(gsiChange, nextState));
        this.addGSI(gsiRecord, tableName, ddbResource);
        this.templateState.add(stackName, JSON.stringify(ddbResource));
      }
      // if its an edit most likely one gsi is removed and another was added
      // by using the index name we can check which values to remove
      else if (gsiStatus === GSIStatus.edit) {
        const gsiPath = gsiChange.path.slice(0, 7);
        const rhsGSIName = _.get(nextState, gsiPath).IndexName;
        const lhsGSIName = _.get(currentState, gsiPath).IndexName;
        // remove the gsi
        this.deleteGSI(lhsGSIName, tableName, ddbResource);
        this.templateState.add(stackName, JSON.stringify(ddbResource));
        // add the gsi
        const gsiRecord = this.getGSIRecord(rhsGSIName, this.getTable(gsiChange, nextState));
        this.addGSI(gsiRecord, tableName, ddbResource);
        this.templateState.add(stackName, JSON.stringify(ddbResource));
      } else if (gsiStatus === GSIStatus.delete) {
        const removedGSI = (gsiChange as any).item.lhs as GlobalSecondaryIndex;
        this.deleteGSI(removedGSI.IndexName as string, tableName, ddbResource);
        this.templateState.add(stackName, JSON.stringify(ddbResource));
      } else if (gsiStatus === GSIStatus.batchAdd) {
        const addedGSIs = (gsiChange as any).lhs as GlobalSecondaryIndex[];
        for (const gsi of addedGSIs) {
          // grab added gsi resources
          let gsiRecord = this.getGSIRecord(gsi.IndexName as string, this.getTable(gsiChange, nextState));
          this.addGSI(gsiRecord, tableName, ddbResource);
          this.templateState.add(stackName, JSON.stringify(ddbResource));
        }
      } else if (gsiStatus === GSIStatus.batchDelete) {
        const removedGSIs = (gsiChange as any).lhs as GlobalSecondaryIndex[];
        for (let gsi of removedGSIs) {
          // grab deleted gsi resource
          this.deleteGSI(gsi.IndexName as string, tableName, ddbResource);
          this.templateState.add(stackName, JSON.stringify(ddbResource));
        }
      }
    }
  };

  private gsiChangeStatus = (gsiChange: Diff<any, any>, current: DiffableProject, next: DiffableProject): GSIStatus => {
    if (gsiChange.kind === 'A') {
      if (gsiChange.item.kind === 'D' && gsiChange.item.lhs) {
        return GSIStatus.delete;
      }
      if (gsiChange.item.kind === 'N' && gsiChange.item.rhs) {
        return GSIStatus.add;
      }
    }
    if (gsiChange.kind === 'E' && gsiChange.lhs) {
      if (gsiChange.path.slice(-1)[0] === 'IndexName') return GSIStatus.edit;
      if (gsiChange.path.slice(-1)[0] === 'AttributeName') {
        // need to run a check to ensure this ks change is actually happening and not because the order changed.
        const innerDiffs = this.getInnerDiffs(gsiChange, current, next);
        const pathToGSI = gsiChange.path.slice(0, 7);
        const gsiIndexName = _.get(current, pathToGSI).IndexName;
        for (const innerDiff of innerDiffs) {
          if (innerDiff.kind === 'E' && innerDiff.path.slice(-1)[0] === 'AttributeName' && innerDiff.path[0] === gsiIndexName) {
            return GSIStatus.edit;
          }
        }
      }
    }
    if (gsiChange.kind === 'N' && gsiChange.rhs.length > 1) {
      return GSIStatus.batchAdd;
    }
    if (gsiChange.kind === 'D' && gsiChange.lhs.length > 1) {
      return GSIStatus.batchDelete;
    }
    return GSIStatus.none;
  };

  private getTable = (gsiChange: Diff<any, any>, proj: DiffableProject): DynamoDB.Table => {
    return proj.stacks[gsiChange.path[1]].Resources[gsiChange.path[3]] as DynamoDB.Table;
  };

  private getStack(stackName: string, proj: DiffableProject): Template {
    return proj.stacks[`${stackName}.json`];
  }

  private getInnerDiffs = (gsiChange: Diff<any, any>, current: DiffableProject, next: DiffableProject) => {
    const pathToGSIs = gsiChange.path.slice(0, 6);
    const oldIndexes = _.get(current, pathToGSIs);
    const newIndexes = _.get(next, pathToGSIs);
    const oldIndexesDiffable = _.keyBy(oldIndexes, 'IndexName');
    const newIndexesDiffable = _.keyBy(newIndexes, 'IndexName');
    return getDiffs(oldIndexesDiffable, newIndexesDiffable) || [];
  };
  /**
   * GSI Operations
   */
  private getGSIRecord = (indexName: string, table: DynamoDB.Table): GSIRecord => {
    const gsis = table.Properties.GlobalSecondaryIndexes as GlobalSecondaryIndex[];
    const addedGSI = (_.filter(gsis, {
      IndexName: indexName,
    }) as GlobalSecondaryIndex[])[0];
    const attrDefs = ((addedGSI.KeySchema as any) as AttributeDefinition[]).reduce((acc, attr) => {
      acc.push(attr.AttributeName);
      return acc;
    }, []);
    const attrDef = _.filter(table.Properties.AttributeDefinitions as AttributeDefinition[], defs => {
      return attrDefs.includes(defs.AttributeName);
    });
    return { gsi: addedGSI, attributeDefinition: attrDef };
  };

  private addGSI = (gsiRecord: GSIRecord, tableName: string, template: Template): void => {
    const table = template.Resources[tableName];
    const gsis = table.Properties.GlobalSecondaryIndexes as GlobalSecondaryIndex[];
    gsis.push(gsiRecord.gsi);
    const attrDefs = table.Properties.AttributeDefinitions as AttributeDefinition[];
    table.Properties.AttributeDefinitions = _.unionBy(attrDefs, gsiRecord.attributeDefinition, 'AttributeName');
  };

  private deleteGSI = (indexName: string, tableName: string, template: Template): void => {
    const table = template.Resources[tableName];
    const gsis = table.Properties.GlobalSecondaryIndexes as GlobalSecondaryIndex[];
    const attrDefs = table.Properties.AttributeDefinitions as AttributeDefinition[];
    const removedGSIKS = _.remove(gsis, { IndexName: indexName })[0]?.KeySchema as Array<KeySchema>;
    const currentKS = gsis.reduce((acc, gsi) => {
      acc.push(...(gsi.KeySchema as Array<KeySchema>));
      return acc;
    }, []);
    // values in removedGSIKS that is not existent in currentKS will be removed
    if (removedGSIKS) {
      const attrToRemove = _.differenceBy(removedGSIKS, currentKS, 'AttributeName');
      _.pullAllBy(attrDefs, attrToRemove, 'AttributeName');
    }
  };
}
