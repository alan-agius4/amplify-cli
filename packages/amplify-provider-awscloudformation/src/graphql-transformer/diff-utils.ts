import { GlobalSecondaryIndex } from 'cloudform-types/types/dynamoDb/table';
import { Diff, diff as getDiffs } from 'deep-diff';
import _ from 'lodash';
import { GSIStatus } from '../utils/amplify-resource-state-utils';
import { DiffableProject } from './utils';

export type IndexChange = {
  type: GSIStatus;
  indexName: string;
};
// Todo rename to something sensible
export const getIndexChanges = (gsiChange: Diff<any, any>, current: DiffableProject, next: DiffableProject): IndexChange[] => {
  const leafPath = gsiChange.path.slice(-1)[0];
  if (gsiChange.kind === 'A') {
    if (gsiChange.item.kind === 'D' && gsiChange.item.lhs) {
      return [
        {
          type: GSIStatus.delete,
          indexName: gsiChange.item.lhs.IndexName,
        },
      ];
    }
    if (gsiChange.item.kind === 'N' && gsiChange.item.rhs) {
      if (['IndexName', 'GlobalSecondaryIndexes'].includes(leafPath)) {
        const innerDiffs = getInnerDiffs(gsiChange, current, next);
        const pathToGSI = gsiChange.path.slice(0, 7);
        const gsiIndexName = _.get(current, pathToGSI).IndexName;
        for (const innerDiff of innerDiffs) {
          if (innerDiff.kind === 'A' && innerDiff.path.slice(-1)[0] === 'KeySchema' && innerDiff.path[0] === gsiIndexName) {
            return [{ type: GSIStatus.edit, indexName: gsiIndexName }];
          }
        }
        return [
          {
            type: GSIStatus.add,
            indexName: gsiChange.item.rhs.IndexName,
          },
        ];
      } else if (['KeySchema'].includes(leafPath)) {
        // ensure the diff is not because the key schema is moved
        const innerDiffs = getInnerDiffs(gsiChange, current, next);
        const pathToGSI = gsiChange.path.slice(0, 7);
        const gsiIndexName = _.get(current, pathToGSI).IndexName;
        for (const innerDiff of innerDiffs) {
          if (innerDiff.kind === 'A' && innerDiff.path.slice(-1)[0] === 'KeySchema' && innerDiff.path[0] === gsiIndexName) {
            return [{ type: GSIStatus.edit, indexName: gsiIndexName }];
          }
        }
      }
    }
  } else if (gsiChange.kind === 'N' && gsiChange.rhs.length >= 1) {
    return gsiChange.rhs.map(newIndex => {
      return {
        type: GSIStatus.add,
        indexName: newIndex.IndexName,
      };
    });
  } else if (gsiChange.kind === 'D' && gsiChange.lhs.length > 1) {
    return gsiChange.lhs.map(newIndex => {
      return {
        type: GSIStatus.delete,
        indexName: newIndex.IndexName,
      };
    });
  } else if (gsiChange.kind === 'E' && gsiChange.lhs) {
    if (leafPath === 'IndexName') {
      const innerDiff = getInnerDiffs(gsiChange, current, next);
      // Index is moved around no real change
      if (innerDiff.length == 0) {
        return [];
      }
      return [
        { type: GSIStatus.delete, indexName: gsiChange.lhs },
        { type: GSIStatus.add, indexName: gsiChange.rhs },
      ];
    }

    if (leafPath === 'AttributeName') {
      // need to run a check to ensure this ks change is actually happening and not because the order changed.
      const innerDiffs = getInnerDiffs(gsiChange, current, next);
      const pathToGSI = gsiChange.path.slice(0, 7);
      const gsiIndexName = _.get(current, pathToGSI).IndexName;
      for (const innerDiff of innerDiffs) {
        if (innerDiff.kind === 'E' && innerDiff.path.slice(-1)[0] === 'AttributeName' && innerDiff.path[0] === gsiIndexName) {
          return [{ type: GSIStatus.edit, indexName: gsiIndexName }];
        }
      }
    }
  }
  return [];
};

export const getInnerDiffs = (gsiChange: Diff<any, any>, current: DiffableProject, next: DiffableProject) => {
  const gsiIndex = gsiChange.path.indexOf('GlobalSecondaryIndexes');
  const pathToGSIs = gsiChange.path.slice(0, gsiIndex + 1);
  const oldIndexes = _.get(current, pathToGSIs);
  const newIndexes = _.get(next, pathToGSIs);
  const oldIndexesDiffable = _.keyBy(oldIndexes, 'IndexName');
  const newIndexesDiffable = _.keyBy(newIndexes, 'IndexName');
  return getDiffs(oldIndexesDiffable, newIndexesDiffable) || [];
};
