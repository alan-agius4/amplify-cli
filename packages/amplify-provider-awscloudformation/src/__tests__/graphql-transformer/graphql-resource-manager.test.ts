import { CloudFormation } from 'aws-sdk';
import { mocked } from 'ts-jest/utils';
import { GraphQLResourceManager } from '../../graphql-transformer/';
import { $ResourceMeta } from '../../graphql-transformer/amplify-graphql-resource-manager';
import { DiffableProject, getGQLDiff } from '../../graphql-transformer/utils';
import { diff as getDiffs } from 'deep-diff';

jest.mock('../../graphql-transformer/utils', () => ({
  ...jest.requireActual('../../graphql-transformer/utils'),
  getGQLDiff: jest.fn(),
}));
jest.mock('../../utils/amplify-resource-state-utils', () => ({
  ...jest.requireActual('../../utils/amplify-resource-state-utils'),
  getTableARNS: jest.fn().mockImplementation(async (cfnClient: CloudFormation, tables: string[], StackId: string) => {
    const map = new Map();
    for (const table of tables) {
      map.set(table, `${table}-${StackId}-dev`);
    }
    return map;
  }),
  getStackParameters: jest.fn().mockImplementation(async (cfnClient: CloudFormation, StackId: string) => {
    return {
      Bucket: `Name-${StackId}`,
      Root: 'sdasdasdasd',
    };
  }),
}));
jest.mock('fs-extra');
jest.mock('../../upload-appsync-files', () => ({
  hashDirectory: jest.fn().mockImplementation((directory: string) => {
    return 'hashbuild';
  }),
}));
jest.mock('aws-sdk');

describe('graphql resource manager', () => {
  const resourceMeta: $ResourceMeta = {
    category: 'api',
    providerPlugin: 'awscloudformation',
    resourceName: 'bookstore',
    service: 'AppSync',
    output: {},
    providerMetadata: {
      s3TemplateURL: 'https://storage.myservice.com/mybookstore/template.json',
      logicalId: 'apibookstore',
    },
    stackId: 'apibookstoreID',
    DeploymentBucketName: 'mybookstore-deployment-bucket',
  };
  test('test with two key change', async () => {
    //JSON.stringify(current.stacks['Book.json'].Resources.BookTable.Properties, null, 2)
    const current: DiffableProject = {
      stacks: {
        'Book.json': {
          Resources: {
            BookTable: {
              Type: 'AWS::DynamoDB::Table',
              Properties: {
                KeySchema: [{ AttributeName: 'id', KeyType: 'HASH' }],
                AttributeDefinitions: [{ AttributeName: 'id', AttributeType: 'S' }],
              },
            },
          },
        },
      },
      root: {},
    };
    const next: DiffableProject = {
      stacks: {
        'Book.json': {
          Resources: {
            BookTable: {
              Type: 'AWS::DynamoDB::Table',
              Properties: {
                KeySchema: [{ AttributeName: 'id', KeyType: 'HASH' }],
                AttributeDefinitions: [
                  { AttributeName: 'id', AttributeType: 'S' },
                  { AttributeName: 'name', AttributeType: 'S' },
                  { AttributeName: 'author', AttributeType: 'S' },
                ],
                GlobalSecondaryIndexes: [
                  { IndexName: 'byName', KeySchema: [{ AttributeName: 'name', KeyType: 'HASH' }] },
                  { IndexName: 'byAuthor', KeySchema: [{ AttributeName: 'author', KeyType: 'HASH' }] },
                ],
              },
            },
          },
        },
      },
      root: {},
    };
    const gqlManager = new GraphQLResourceManager({
      cfnClient: (jest.fn() as unknown) as CloudFormation,
      resourceMeta: resourceMeta,
      backendDir: 'localDir',
      cloudBackendDir: 'cloudDir',
    });
    mocked(getGQLDiff).mockImplementation((currentBacked: string, cloudBackendDir: string) => {
      return { current, next, diff: getDiffs(current, next) };
    });
    const deploymentSteps = await gqlManager.run();
    expect(deploymentSteps).toBeDefined();
    expect(deploymentSteps).toMatchSnapshot();
  });
});
