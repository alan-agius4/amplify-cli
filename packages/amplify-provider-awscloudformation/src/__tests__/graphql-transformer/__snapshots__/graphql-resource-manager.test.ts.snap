// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql resource manager test with two key change 1`] = `
Array [
  Object {
    "parameters": Object {
      "Bucket": "Name-apibookstoreID",
      "Root": "sdasdasdasd",
      "S3DeploymentRootKey": "undefined/hashbuild/states/0",
    },
    "stackName": "apibookstoreID",
    "stackTemplatePath": "https://storage.myservice.com/mybookstore/template.json",
    "tableNames": Array [
      "Book-apibookstoreID-dev",
    ],
  },
  Object {
    "parameters": Object {
      "Bucket": "Name-apibookstoreID",
      "Root": "sdasdasdasd",
      "S3DeploymentRootKey": "undefined/hashbuild/states/1",
    },
    "stackName": "apibookstoreID",
    "stackTemplatePath": "https://storage.myservice.com/mybookstore/template.json",
    "tableNames": Array [
      "Book-apibookstoreID-dev",
    ],
  },
]
`;
